OSDOCS#1524 - Update OVN-Kubernetes migration procedure (#26102)
openshift/openshift-docs: modules/nw-ovn-kubernetes-migration.adoc

+
[source,terminal]
----
$ oc patch Network.operator.openshift.io cluster --type=merge \
  --patch '{
    "spec":{
      "defaultNetwork":{
        "ovnKubernetesConfig":{
          "mtu":<mtu>,
          "genevePort":<port>
    }}}}'
----
+
where:
+
--
`mtu`::
The MTU for the Geneve overlay network. This value is normally configured automatically, but if the nodes in your cluster do not all use the same MTU, then you must set this explicitly to `100` less than the smallest node MTU value.
`port`::
The UDP port for the Geneve overlay network.
--
+
.Example patch command 
[source,terminal]
----
$ oc patch Network.operator.openshift.io cluster --type=merge \
  --patch '{
    "spec":{
      "defaultNetwork":{
        "ovnKubernetesConfig":{
          "mtu":1200
    }}}}'
----

. Wait until the Multus daemon set pods restart. The name of the Multus pods is in form of `multus-<xxxxx>` where `<xxxxx>` is a random sequence of letters. It might take several moments for the pods to restart.
+
[source,terminal]
----

> $ watch oc get pod -n openshift-multus

Reviewer: Suggest use this way to check the multus pods rollout.  'oc -n openshift-multus rollout status daemonset/multus'

Author: @huiran0826, what output does that command produce? Thanks!

Reviewer: + oc -n openshift-multus rollout status daemonset/multus
Waiting for daemon set "multus" rollout to finish: 1 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 1 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 1 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 2 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 2 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 2 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 2 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 3 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 3 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 3 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 3 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 4 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 4 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 4 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 4 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 5 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 5 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 5 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 5 out of 6 new pods have been updated...
Waiting for daemon set "multus" rollout to finish: 5 of 6 updated pods are available...
daemon set "multus" successfully rolled out

----

. Configure the OVN-Kubernetes cluster network provider with one of the following commands:

.. To specify the network provider without changing the cluster network IP address block, enter the following command:
+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster \
  --type='merge' --patch '{ "spec": { "networkType": "OVNKubernetes" } }'
----

.. To specify a different cluster network IP address block, enter the following command:
+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster \
  --type='merge' --patch '{
    "spec": {
      "clusterNetwork": [
        "cidr": "<cidr>",
        "hostPrefix": "<prefix>"
      ]
      "networkType": "OVNKubernetes"
    } 
  }'
----
+
where `cidr` is a CIDR block and `prefix` is the slice of the CIDR block apportioned to each node in your cluster.
+
IMPORTANT: You cannot change the service network address block during the migration.

. Optional: You can customize the following settings for OVN-Kubernetes to meet your network infrastructure requirements:
+
--
* Maximum transmission unit (MTU)
* Geneve (Generic Network Virtualization Encapsulation) overlay network port
--
+

> Enter the following command to open the Cluster Network Operator (CNO) configuration in an editor:

Reviewer 2: Users shall update MTU and Geneve port within the same patch command of changing the `NetworkType`. Users can choose to use either `oc patch` or `oc edit` with one transaction.  But not both of them, neither in separate transactions.

Author: @pliurh, I'm confused. I do not see any fields for MTU or Geneve in the `Network.config` CRD.

https://docs.openshift.com/container-platform/4.6/rest_api/config_apis/network-config-openshift-io-v1.html#network-config-openshift-io-v1

How to patch `Network.config` and `Network.operator` in the same command?

Thanks.

https://docs.openshift.com/container-platform/4.6/rest_api/operator_apis/network-operator-openshift-io-v1.html#network-operator-openshift-io-v1

Reviewer 2: Oh, my fault. Please ignore that comment.

Reviewer 2: But still, users shall execute these 2 commands together, so that the CNO can rendor the ovn-kube object correctly before applying it. So using 'oc patch' is more handy here, e.g. `oc patch Network.operator.openshift.io cluster --type='merge' --patch '{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"mtu":1200}}}}'`

Author: @pliurh, I've made another update to this.

Reviewer 2: There is a bug https://bugzilla.redhat.com/show_bug.cgi?id=1898159, which crashes the cluster if users do migrate with a new clusterNetwork CIDR and the new CIDR doesn't include the old CIDR. Shall we remove this session until that bug is closed?

Author: Sure, I've moved the offending text into a comment at the bottom of the module.

  --type='merge' --patch '{ "spec": { "networkType": "OVNKubernetes" } }'
----

> . Optional: You can customize the maximum transmission unit (MTU) and the port used by the Geneve (Generic Network Virtualization Encapsulation) overlay network to meet your network infrastructure requirements. Enter the following command to open the Cluster Network Operator (CNO) configuration in an editor:

Author: @pliurh, this adds an optional step for configuring the CNO with a different MTU or Geneve port.

Reviewer 2: Users can also change the clusterNetwork CIDR during the migration. But  the serviceNetwork shall not be changed.

Reviewer 2: Changing the Network.config should be put in one step.

Author: @pliurh, okay, that makes sense.

> +

Reviewer 2: line 63 shall also be updated. Since we will watch the pods of multus, not openshift-sdn.

[source,terminal]
----

> $ watch oc get pod -n openshift-multus

Reviewer: This part still need confirmation from @pliurh ,according to the new procedure doc,after update the network type,"Check there is no pod under namespace ‘openshift-multus’", however the pods multus-XXXX just got recreated and still in running status.  And other pods in namespace openshift-multus like multus-admission-controller-XXXXX and network-metrics-daemon-XXXXX were not recreated.

Reviewer 2: It should be 'wait until all the pods of the DaemonSet `multus` are recreated'.

Author: @pliurh @huiran0826 

Is there a specific pattern that a user should look for, like `multus-XXXX`, or is it all pods in the `openshift-multus` namespace?

Reviewer 2: `multus-XXXX`

Author: I've updated to include:

> The name of the Multus pods is in form of `multus-<xxxx>` where `<xxxx>` is a random sequence of letters.

done
----

. After the nodes in your cluster have rebooted, start all of the machine configuration pools:
+
[source,terminal]
----
$ oc patch MachineConfigPool master --type='merge' --patch \
  '{ "spec": { "paused": false } }'
----
+
[source,terminal]
----
$ oc patch MachineConfigPool worker --type='merge' --patch \
  '{ "spec": { "paused": false } }'
----

. To confirm the status of the new machine configuration on a host, enter the following command. Replace `<node>` with the name of the node.
+
[source,terminal]
----

> $ oc describe <node> | egrep "hostname|machineconfig"

Reviewer: Just use   oc describe node | egrep "hostname|machineconfig",  it will list all the nodes status.

----
+
When the new configuration is applied, the value of `machineconfiguration.openshift.io/state` field is `Done`.

> 

Reviewer 2: Add a note here, the MCO will reboot every node to apply the new machine config. By default, the MCO will do it one at a time for each machine pool. So the time of this whole migration will grow with the size of the cluster.

done
----

> . After the nodes in your cluster have rebooted, start all of the machine configuration pools. As the MCO updates machines in each config pool, by default it updates a single machine per pool at a time, so the time that the migration requires to complete grows with the size of the cluster.

Reviewer 2: To apply the new MachineConfig to a node, the MCO will trigger a reboot for each node.

Author: @pliurh updated to

> After the nodes in your cluster have rebooted, start all of the machine configuration pools. As the MCO updates machines in each config pool, it reboots each node. By default the MCO updates a single machine per pool at a time, so the time that the migration requires to complete grows with the size of the cluster.

Reviewer 2: Here comes the output example of `oc describe node | egrep "hostname|machineconfig"` for one node.

```
                    kubernetes.io/hostname=master-0
                    machineconfiguration.openshift.io/currentConfig: rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b
                    machineconfiguration.openshift.io/desiredConfig: rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b
                    machineconfiguration.openshift.io/reason:
                    machineconfiguration.openshift.io/state: Done
```

When the `state` is done, the `currentConfig` shall equal to the `desiredConfig`. The name of the MachineConfig can be found by `oc get machineconfig`. When we trigger the migration, a new MachineConfig will be rendered by MCO for each MachineConfigPool. Users have to make sure that the newly rendered MachineConfig has been applied to the node successfully by comparing the name of `currentConfig` and the output of `oc get machineconfig`.

```
[root@nfvpe-21 ~]# oc get mc
...
rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b    6c8bd9b2915c44e03ec10e02e59fedb07e3807cb   3.1.0             6d23h
...
```

+
[source,terminal]
----
$ oc patch MachineConfigPool master --type='merge' --patch \
  '{ "spec": { "paused": false } }'
----
+
[source,terminal]
----
$ oc patch MachineConfigPool worker --type='merge' --patch \
  '{ "spec": { "paused": false } }'
----

. To confirm the status of the new machine configuration on the hosts, enter the following command.
+
[source,terminal]
----
$ oc describe node | egrep "hostname|machineconfig"
----
+

> When the new configuration is applied, the value of `machineconfiguration.openshift.io/state` field is `Done`.

Reviewer 2: Users shall check whether the machineconfiguration.openshift.io/currentConfig of all the nodes from one machine config pool is the same and is the name of the latest MachineConfig.

Author: @pliurh is this in addition to `machineconfiguration.openshift.io/state`?

How can someone do this? Is there a set of commands I can include for this verification step?

Thanks.

Reviewer 2: Here comes the output example of `oc describe node | egrep "hostname|machineconfig"` for one node.

```
                    kubernetes.io/hostname=master-0
                    machineconfiguration.openshift.io/currentConfig: rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b
                    machineconfiguration.openshift.io/desiredConfig: rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b
                    machineconfiguration.openshift.io/reason:
                    machineconfiguration.openshift.io/state: Done
```

When the `state` is done, the `currentConfig` shall equal to the `desiredConfig`. The name of the MachineConfig can be found by `oc get machineconfig`. When we trigger the migration, a new MachineConfig will be rendered by MCO for each MachineConfigPool. Users have to make sure that the newly rendered MachineConfig has been applied to the node successfully by comparing the name of `currentConfig` and the output of `oc get machineconfig`.

```
[root@nfvpe-21 ~]# oc get mc
...
rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b    6c8bd9b2915c44e03ec10e02e59fedb07e3807cb   3.1.0             6d23h
...
```

Author: @pliurh, the challenge with MCO is that the `rendered-master-...` hash is mostly meaningless. If there are a dozen of these in the list, how does someone know that `c53e221d9d24e1c8bb6ee89dd3d8ad7b` is what must be in `machineconfiguration.openshift.io/currentConfig`?

If it is the most recently created resource, that might be a good guess. But it isn't definitive.

Is it possible to sort `oc get mc` by creation time?

Does `oc get mc rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b -o yaml` provide any useful information? Is it possible to see if `rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b` is correct some other way than listing all machineconfig resources?

Reviewer 2: I always use the creation time as an indicator.  However, `oc get mc rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b -o yaml` can help as well. In the ignition config of the machine config CR, user can check the systemd unit of `ovs-configuration`. If it's for ovn-kubernetes, you can see `ExecStart=/usr/local/bin/configure-ovs.sh OVNKubernetes`; for openshift-sdn, you can see `ExecStart=/usr/local/bin/configure-ovs.sh OpenShiftSDN`

Author: @pliurh, do you have any example command output for:

`oc get mc rendered-master-c53e221d9d24e1c8bb6ee89dd3d8ad7b -o yaml`

That I can include in the document?

Thanks!

----
$ oc get nodes
----

> .. If a node is stuck in the `NotReady` state, investigate the `machine-config-daemon-<seq>` pod logs and resolve any errors. The `<seq>` value is a random five character alphanumeric sequence.

Reviewer 3: I'd move the information about `<seq>` to after the example output. It's good detail, but I don't think it's critical.

... To list the pods, enter the following command:
+
[source,terminal]
----
$ oc get pod -n openshift-machine-config-operator
----
+
.Example output
[source,terminal]
----
NAME                                         READY   STATUS    RESTARTS   AGE
machine-config-controller-75f756f89d-sjp8b   1/1     Running   0          37m
machine-config-daemon-5cf4b                  2/2     Running   0          43h
machine-config-daemon-7wzcd                  2/2     Running   0          43h
machine-config-daemon-fc946                  2/2     Running   0          43h
machine-config-daemon-g2v28                  2/2     Running   0          43h
machine-config-daemon-gcl4f                  2/2     Running   0          43h
machine-config-daemon-l5tnv                  2/2     Running   0          43h
machine-config-operator-79d9c55d5-hth92      1/1     Running   0          37m
machine-config-server-bsc8h                  1/1     Running   0          43h
machine-config-server-hklrm                  1/1     Running   0          43h
machine-config-server-k9rtx                  1/1     Running   0          43h
----

> ... To display the pod log for each machine config daemon pod shown in the previous output, enter the following command:

Reviewer 3: I don't think it's quite clear enough that you need to repeat these two steps for each pod.
```suggestion
... Display the pod log for the first machine config daemon pod shown in the previous output by enter the following command:
```
Then add a step after line 241 that says something like "Repeat the previous two steps for each machine config daemon pod."

----
$ oc delete namespace openshift-sdn
----

// https://bugzilla.redhat.com/show_bug.cgi?id=1898159
////
. Configure the OVN-Kubernetes cluster network provider with one of the following commands:

> .. To specify the network provider without changing the cluster network IP address block, enter the following command:

Reviewer 3: ```suggestion
** To specify the network provider without changing the cluster network IP address block, enter the following command:
```

Since you only do one, I'd make this a UL.
